### PR TITLE
#571 NativeImageLoader buffer reuse

### PR DESCRIPTION
**Pull request by AlexDBlack**
***WIP DO NOT MERGE***
Fixes: https://github.com/deeplearning4j/DataVec/issues/571

All tests pass, good to go except for loading issue noted below (which should be resolved first).

@saudet without adding an Nd4j.create(1) at the top of the test (and 1 or 2 others now) I'm getting the following (when the tests are run in isolation):
```
java.lang.RuntimeException: No native JavaCPP library in memory. (Has Loader.load() been called?)
	at org.bytedeco.javacpp.BytePointer.<init>(BytePointer.java:103)
	at org.datavec.image.loader.NativeImageLoader.asMatrix(NativeImageLoader.java:211)
	at org.datavec.image.loader.NativeImageLoader.asMatrix(NativeImageLoader.java:202)
	at org.datavec.image.loader.TestNativeImageLoader.testBufferRealloc(TestNativeImageLoader.java:330)
...
Caused by: java.lang.UnsatisfiedLinkError: org.bytedeco.javacpp.BytePointer.allocateArray(J)V
	at org.bytedeco.javacpp.BytePointer.allocateArray(Native Method)
	at org.bytedeco.javacpp.BytePointer.<init>(BytePointer.java:98)
	... 25 more
```

What's the best workaround/solution here?

### CODE REVIEW

File: datavec-data/datavec-data-image/src/main/java/org/datavec/image/loader/NativeImageLoader.java [hunk 1]
```diff
@@ -205,10 +205,16 @@ public INDArray asMatrix(File f) throws IOException {
...
     @Override
     public INDArray asMatrix(InputStream is) throws IOException {
-        byte[] bytes = IOUtils.toByteArray(is);
-        Mat image = imdecode(new Mat(bytes), CV_LOAD_IMAGE_ANYDEPTH | CV_LOAD_IMAGE_ANYCOLOR);
+        Pair<byte[],Integer> pair = streamToBuffer(is);
+        byte[] bytes = pair.getFirst();
+        int length = pair.getSecond();
+        BytePointer bp = new BytePointer(length);
+        bp.put(bytes, 0, length);
+        Mat mat = new Mat(bp, false);
```
Comment (saudet) [inline datavec-data/datavec-data-image/src/main/java/org/datavec/image/loader/NativeImageLoader.java]: We don't need to create a BytePointer explicitly here, just call `new Mat(bytes)` like before and it *should* work.

BTW, creating a new BytePointer like that on each call is also bound to create a lot of garbage. You're not seeing that?

Comment (saudet) [reply datavec-data/datavec-data-image/src/main/java/org/datavec/image/loader/NativeImageLoader.java]: Let me take a stab at it

**Comment by AlexDBlack** [comment]
@saudet Any progress here? Or should we merge this as is (to reduce the byte[] memory use) and look at the pointer/offheap component separately?

**Comment by saudet** [comment]
Ah yes let me do this now actually.

**Pull request merged**